core/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java

```diff
@@ -77,15 +73,35 @@ private void analyze(ActivityPayload payload, UserAnonymizedDTO userAnonymized,
 
 	private void addOrUpdateActivity(ActivityPayload payload, UserAnonymizedDTO userAnonymized, Goal matchingGoal)
 	{
-		Date minEndTime = new Date(payload.startTime.getTime() - yonaProperties.getAnalysisService().getConflictInterval());
-		DayActivity dayActivity = cacheService.fetchDayActivityForUser(userAnonymized.getID(), matchingGoal.getID());
-		if (dayActivity != null && dayActivity.getStartTime().isBefore(getStartOfDay(payload.startTime, userAnonymized)))
+		ActivityPayload truncatedPayload = payload;
+		ActivityPayload nextDayPayload = null;
+		if (isCrossDayActivity(payload, userAnonymized))
 		{
-			// Last day cached was not today
-			dayActivity = null;
+			// crossing day border
+			// set current payload to end of day
+			// use new payload for remainder
+			// assumption: activity never crosses 2 days
+			truncatedPayload = new ActivityPayload(payload.url, payload.startTime,
+					Date.from(getEndOfDay(payload.startTime, userAnonymized).toInstant()), payload.application);
+
+			nextDayPayload = new ActivityPayload(payload.url,
+					Date.from(getStartOfDay(payload.endTime, userAnonymized).toInstant()), payload.endTime, payload.application);
 		}
-		Activity activity = dayActivity == null ? null : dayActivity.getLastActivity();
 
+		addOrUpdateDayTruncatedActivity(truncatedPayload, userAnonymized, matchingGoal);
+
+		if (nextDayPayload != null)
+		{
+			// day should be created
+			addActivity(nextDayPayload, userAnonymized, matchingGoal, null);
+		}
+	}
+
+	private void addOrUpdateDayTruncatedActivity(ActivityPayload payload, UserAnonymizedDTO userAnonymized, Goal matchingGoal)
```

Comment:
Propose to refactor this method to move the nitty gritty details to separate methods:
```
	private void addOrUpdateDayTruncatedActivity(ActivityPayload payload, UserAnonymizedDTO userAnonymized, Goal matchingGoal)
	{
		DayActivity dayActivity = getCachedDayActivity(payload, userAnonymized, matchingGoal);
		Activity lastRegisteredActivity = dayActivity == null ? null : dayActivity.getLastActivity();
		if (isInIntervalWithLastRegisteredActivity(lastRegisteredActivity, payload))
		{
			addActivity(payload, userAnonymized, matchingGoal, dayActivity);
		}
		else if (isBeyondSkipWindowAfterLastRegisteredActivity(lastRegisteredActivity, payload))
		{
			updateActivity(payload, userAnonymized, matchingGoal, dayActivity, lastRegisteredActivity);
		}
	}

	private boolean isBeyondSkipWindowAfterLastRegisteredActivity(Activity lastRegisteredActivity, ActivityPayload payload)
	{
		return payload.endTime.getTime() - lastRegisteredActivity.getEndTime().getTime() >= yonaProperties.getAnalysisService()
				.getUpdateSkipWindow() || payload.startTime.before(lastRegisteredActivity.getStartTime());
	}

	private boolean isInIntervalWithLastRegisteredActivity(Activity lastRegisteredActivity, ActivityPayload payload)
	{
		Date minEndTime = new Date(payload.startTime.getTime() - yonaProperties.getAnalysisService().getConflictInterval());
		return lastRegisteredActivity == null || lastRegisteredActivity.getEndTime().before(minEndTime);
	}
```